File: lib/java/src/main/java/com/workiva/frugal/transport/FBoundedMemoryBuffer.java
package com.workiva.frugal.transport;

import com.workiva.frugal.exception.FMessageSizeException;
import org.apache.thrift.transport.TMemoryBuffer;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

/**
 * An implementation of TTransport using a bounded memory buffer. Writes which cause the buffer to exceed its size
 * throw an FMessageSizeException.
 */
public class FBoundedMemoryBuffer extends TTransport {

    private TMemoryBuffer buffer;
    private final int limit;

    public FBoundedMemoryBuffer(int size) {
        buffer = new TMemoryBuffer(size);
        limit = size;
    }

    @Override
    public boolean isOpen() {
        return buffer.isOpen();
    }

    @Override
    public void open() throws TTransportException {
        buffer.open();
    }

    @Override
    public void close() {
        buffer.close();
    }

    @Override
    public int read(byte[] buf, int off, int len) throws TTransportException {
        return buffer.read(buf, off, len);
    }

    @Override
    public void write(byte[] buf) throws TTransportException {
        if (buffer.length() + buf.length > limit) {
            buffer = new TMemoryBuffer(limit);
            throw new FMessageSizeException(String.format("Buffer size reached (%d)", limit));
        }
        buffer.write(buf);
    }

    @Override
    public void write(byte[] buf, int off, int len) throws TTransportException {
        if (buffer.length() + len > limit) {
            buffer = new TMemoryBuffer(limit);
            throw new FMessageSizeException(String.format("Buffer size reached (%d)", limit));
        }
        buffer.write(buf, off, len);
    }

    public int length() {
        return buffer.length();
    }

    public byte[] getArray() {
        return buffer.getArray();
    }

}

File: lib/python/frugal/tornado/transport/f_bounded_memory_buffer.py
from frugal.exceptions import FMessageSizeException
from thrift.transport.TTransport import TMemoryBuffer

_NATS_MAX_MESSAGE_SIZE = 1024 * 1024


class FBoundedMemoryBuffer(TMemoryBuffer, object):
    """An implementation of TMemoryBuffer using a bounded memory buffer.
    Writes which cause the buffer to exceed its size throw an FMessageSizeException"""

    def __init__(self, value=None):
        super(FBoundedMemoryBuffer, self).__init__(value)

    def write(self, buf):
        """Bounded write to buffer"""
        if len(self) + len(buf) > _NATS_MAX_MESSAGE_SIZE:
            self._buffer = TMemoryBuffer()
            raise FMessageSizeException("Buffer size reached {}".format(_NATS_MAX_MESSAGE_SIZE))
        self._buffer.write(buf)

    def __len__(self):
        return len(self.getvalue())
